dlib/dnn/core.h

```diff
@@ -1430,6 +1436,11 @@ namespace dlib
 
         unsigned int sample_expansion_factor() const { return _sample_expansion_factor; }
 
+        void zero_out_gradients()
+        {
+            x_grad = 0;
```

Comment:
```suggestion
            gradient_input_is_stale = true;
```
This should do it and avoid unnecessary assignments when it's already 0.